naff/client/auto_shard_client.py

```diff
@@ -0,0 +1,230 @@
+                    break
+                self._guild_event.clear()
+                if all(self.cache.get_guild(g_id) is not None for g_id in expected_guilds):
+                    # all guilds cached
+                    break
+
+            if self.fetch_members:
+                log.info(f"Shard {shard_id} is waiting for members to be chunked")
+                await asyncio.gather(*(guild.chunked.wait() for guild in self.guilds if guild.id in expected_guilds))
+        else:
+            log.warning(
+                f"Shard {shard_id} reports it has 0 guilds, this is an indicator you may be using too many shards"
+            )
+        # noinspection PyProtectedMember
+        connection_state._shard_ready.set()
+        log.debug(f"Shard {shard_id} is now ready")
+        # noinspection PyProtectedMember
+        await asyncio.gather(*[shard._shard_ready.wait() for shard in self._connection_states])
+
+        # run any pending startup tasks
+        if self.async_startup_tasks:
+            try:
+                await asyncio.gather(*self.async_startup_tasks)
+            except Exception as e:
+                await self.on_error("async-extension-loader", e)
+
+        # cache slash commands
+        if not self._startup:
+            await self._init_interactions()
+
+        if not self._ready.is_set():
+            self._ready.set()
+            if not self._startup:
+                self._startup = True
+                self.dispatch(events.Startup())
+            self.dispatch(events.Ready())
+
+    async def astart(self, token) -> None:
+        """
+        Asynchronous method to start the bot.
+
+        Args:
+            token: Your bot's token
+        """
+        log.debug("Starting http client...")
+        await self.login(token)
+
+        tasks = []
+
+        # Sort shards into their respective ratelimit buckets
+        shard_buckets = defaultdict(list)
+        for shard in self._connection_states:
+            bucket = str(shard.shard_id % self.max_start_concurrency)
+            shard_buckets[bucket].append(shard)
+
+        for bucket in shard_buckets.values():
+            for shard in bucket:
+                log.debug(f"Starting {shard.shard_id}")
+                start = time.perf_counter()
+                tasks.append(asyncio.create_task(shard.start()))
+
+                if self.max_start_concurrency == 1:
+                    # connection ratelimiting when discord has asked for one connection concurrently
+                    # we could wait for `state._shard_ready`, but by waiting on the raw websocket ready
+                    # we can start our subsequent shards sooner - speeding up startup
+                    while not shard.gateway:
+                        await asyncio.sleep(0.1)
+                    # noinspection PyProtectedMember
+                    await shard.gateway._ready.wait()
+                    await asyncio.sleep(5.1 - (time.perf_counter() - start))
```

Comment:
What is this arbitrary amount of sleep?

Reply:
Its not arbitrary

Discord requires 5 seconds between subsiquent batches of connections. Through testing i found this drifted by about 0.1ms, likely down to asyncio imprecision and discord's servers. 5.1 seconds allowed the bot to successfully connect without being hit by discord invalidating the session.